Add VK_KHR_pipeline_binary sample

gpx1000 opened the pull request:
## Description

Introduce a new Vulkan sample demonstrating VK_KHR_pipeline_binary usage. It includes querying pipeline keys, capturing pipeline binaries, and managing them explicitly. The sample is accompanied by documentation, CMake lists, shaders, and updates to the Vulkan framework mapping.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [ ] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [x] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [x] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

gpx1000 committed:
Add VK_KHR_pipeline_binary sample and related resources

Introduce a new Vulkan sample demonstrating VK_KHR_pipeline_binary usage. It includes querying pipeline keys, capturing pipeline binaries, and managing them explicitly. The sample is accompanied by documentation, CMake lists, shaders, and updates to the Vulkan framework mapping.

gpx1000 committed:
update the clang format and copyright.

gpx1000 committed:
update nav.adoc

SaschaWillems commented on samples/extensions/pipeline_binary/README.adoc:
```diff
@@ -0,0 +1,78 @@
+- See the License for the specific language governing permissions and
+- limitations under the License.
+-
+////
```
This is missing the "Where to get the source" Antora block (see readme.adoc for other samples or the readme from the samples template).

SaschaWillems commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,273 @@
+	vkGetPhysicalDeviceFeatures2(get_device().get_gpu().get_handle(), &features2);
+	vkGetPhysicalDeviceProperties2(get_device().get_gpu().get_handle(), &props2);
+
+	LOGI("VK_KHR_pipeline_binary supported feature: pipelineBinaries = {}", features.pipelineBinaries ? "true" : "false");
```
This information should be display in the applications GUI instead of simply outputting it to the terminal/log.

SaschaWillems commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,273 @@
+	res = vkGetPipelineBinaryDataKHR(get_device().get_handle(), &binary_info, &key2, &binary_size, data.data());
+	if (res == VK_SUCCESS)
+	{
+		LOGI("Retrieved pipeline binary of {} bytes; key size {} bytes", static_cast<uint32_t>(binary_size), key2.keySize);
```
Same as above. Should be visible in the GUI instead.

SaschaWillems commented on samples/extensions/pipeline_binary/README.adoc:
```diff
@@ -0,0 +1,78 @@
+- Not all implementations will return data; some may return zero bytes. Handle that gracefully.
+- Any change in creation parameters changes the key and the resulting binary.
+
+== References
```
Should have links to the references, preferably as xrefs inside the docs site.

gpx1000 committed:
- Enhance README with Antora site-gen support and improved references.
- Add detailed logging for pipeline binary operations.
- Introduce UI overlay in `PipelineBinary` to display aggregated log information.
- General code clean-up and formatting adjustments.

gpx1000 committed:
Adjust code formatting in pipeline_binary.cpp for improved readability

gary-sweet reviewed (commented):
I can only really say that this builds ok as we don't currently support pipeline binaries.

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,319 @@
+	LOGI("VK_KHR_pipeline_binary supported feature: pipelineBinaries = {}", pb);
+	char buf[256];
+	snprintf(buf, sizeof(buf), "VK_KHR_pipeline_binary supported feature: pipelineBinaries = %s\n", pb);
+	log_text_ += buf;
```
Instead of all those fixed-size buffers, you could as well use std::format:
```
std::string message = std::format( "VK_KHR_pipeline_binary supported feature: pipelineBinaries = {}", pb );
```
or maybe
```
std::string message = std::format("VK_KHR_pipeline_binary supported feature: pipelineBinaries = {}", static_cast<bool>(features.pipelineBinaries));
```

asuessenbach replied:
Or:
```
	std::string message = std::format("VK_KHR_pipeline_binary support: pipelineBinaries = {}", !!features.pipelineBinaries);
	LOGI(message);
	log_text_ += message + "\n";
```

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,319 @@
+	const char *pic  = props.pipelineBinaryPrefersInternalCache ? "true" : "false";
+	const char *pic2 = props.pipelineBinaryPrecompiledInternalCache ? "true" : "false";
+	const char *cd   = props.pipelineBinaryCompressedData ? "true" : "false";
+	LOGI("VK_KHR_pipeline_binary properties: internalCache={}, internalCacheControl={}, prefersInternalCache={}, precompiledInternalCache={}, compressedData={}", ic, icc, pic, pic2, cd);
```
Maybe add some newlines and tabs to make the output more readable.

asuessenbach replied:
And using `std::format`:
```
  message = std::format(
	    "VK_KHR_pipeline_binary properties: internalCache={}, internalCacheControl={}, prefersInternalCache={}, precompiledInternalCache={}, compressedData={}",
	    !!props.pipelineBinaryInternalCache,
	    !!props.pipelineBinaryInternalCacheControl,
	    !!props.pipelineBinaryPrefersInternalCache,
	    !!props.pipelineBinaryPrecompiledInternalCache,
	    !!props.pipelineBinaryCompressedData);
	LOGI(message);
  log_text_ += message + "\n";
```

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,319 @@
+	}
+}
+
+void PipelineBinary::on_update_ui_overlay(vkb::Drawer &drawer)
```
Despite overriding `on_update_ui_overlay`, I don't see any UI.
Don't know why.

asuessenbach reviewed (commented):
Nice little sample. Just a few comments...

ShabbyX commented on samples/extensions/pipeline_binary/README.adoc:
```diff
@@ -0,0 +1,82 @@
+- VK_EXT_graphics_pipeline_library reduces link-time work, but doesn’t hand you the final binary for persistence.
+- VK_EXT_shader_object changes the binding model, not binary persistence.
+
+== How it works (TL;DR)
```
You could consider linking to https://www.khronos.org/blog/bringing-explicit-pipeline-caching-control-to-vulkan somewhere, which has some graphs that could help understand these better.

gpx1000 committed:
Merge branch 'main' into VK_KHR_pipeline_binaries

# Conflicts:
#	framework/vulkan_type_mapping.h
#	samples/extensions/README.adoc

gpx1000 committed:
Update pipeline binary sample with interactive demo and performance metrics

Add interactive GUI to demonstrate pipeline recreation from scratch vs. from cached binaries with timing measurements. Implement binary persistence (save/load to disk) and display performance statistics including speedup calculations. Update documentation to reflect new interactive features and workflow.

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,613 @@
+		LOGW("vkGetPipelineKeyKHR failed ({}); skipping binary capture", static_cast<int>(res));
+		char buf[128];
+		snprintf(buf, sizeof(buf), "vkGetPipelineKeyKHR failed (%d); skipping binary capture\n", static_cast<int>(res));
+		log_text_ += buf;
```
using `std::format` (and `vk::to_string`):
```
		std::string message = std::format( "vkGetPipelineKeyKHR failed ({}); skipping binary capture", vk::to_string(static_cast<vk::Result>(res)) );
		LOGW(message);
		log_text_ += message + "\n";
```

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,613 @@
+		char buf[128];
+		snprintf(buf, sizeof(buf), "Got pipeline key (%u bytes)\n", static_cast<unsigned>(key.keySize));
+		log_text_ += buf;
+	}
```
With `std::format`:
```
	std::string message = std::format("Got pipeline key ({} bytes)", key.keySize);
	LOGI(message);
	log_text_ += message + "\n";
```

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,613 @@
+		LOGW("vkCreatePipelineBinariesKHR failed ({}); driver may not support capturing binaries in this context", static_cast<int>(res));
+		char buf[180];
+		snprintf(buf, sizeof(buf), "vkCreatePipelineBinariesKHR failed (%d); driver may not support capturing binaries in this context\n", static_cast<int>(res));
+		log_text_ += buf;
```
With `std::format`:
```
		std::string message = std::format("vkCreatePipelineBinariesKHR failed ({}); driver may not support capturing binaries in this context",
		                                  vk::to_string(static_cast<vk::Result>(res)));
		LOGW(message);
		log_text_ += message + "\n";
```

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,613 @@
+		LOGW("vkGetPipelineBinaryDataKHR size query failed ({}); skipping data fetch", static_cast<int>(res));
+		char buf[160];
+		snprintf(buf, sizeof(buf), "vkGetPipelineBinaryDataKHR size query failed (%d); skipping data fetch\n", static_cast<int>(res));
+		log_text_ += buf;
```
With `std::format`:
```
		std::string message = std::format("vkGetPipelineBinaryDataKHR size query failed ({}); skipping data fetch", vk::to_string(static_cast<vk::Result>(res)));
		LOGW(message);
		log_text_ += message + "\n";
```

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,613 @@
+			char sig[96];
+			snprintf(sig, sizeof(sig), "Binary signature: %u %u %u %u ...\n", binary_data_[0], binary_data_[1], binary_data_[2], binary_data_[3]);
+			log_text_ += sig;
+			LOGD("Binary signature: {} {} {} {} ...", binary_data_[0], binary_data_[1], binary_data_[2], binary_data_[3]);
```
With `std::format`:
```
			std::string message = std::format("Binary signature: {} {} {} {} ...", binary_data_[0], binary_data_[1], binary_data_[2], binary_data_[3]);
			LOGD(message);
			log_text_ += message + "\n";
```

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -0,0 +1,613 @@
+		LOGW("vkGetPipelineBinaryDataKHR failed ({}); data not available", static_cast<int>(res));
+		char buf[128];
+		snprintf(buf, sizeof(buf), "vkGetPipelineBinaryDataKHR failed (%d); data not available\n", static_cast<int>(res));
+		log_text_ += buf;
```
With `std::format`:
```
		std::string message = std::format("vkGetPipelineBinaryDataKHR failed ({}); data not available", vk::to_string(static_cast<vk::Result>(res)));
		LOGW(message);
		log_text_ += message + "\n";
```

asuessenbach commented on third_party/volk:
Why do you need a new volk version?
Shouldn't that be done by a separate PR?

gpx1000 committed:
Refactor logging to use std::format and eliminate snprintf

Replace C-style string formatting (snprintf with char buffers) with std::format for cleaner, more modern C++ code. Use vk::to_string for VkResult enum values instead of casting to int. Consolidate duplicate LOGI/LOGW calls and log_text_ updates into single operations per message.

asuessenbach commented on samples/extensions/pipeline_binary/pipeline_binary.cpp:
```diff
@@ -222,18 +221,17 @@ void PipelineBinary::demo_pipeline_key_and_binary()
-		snprintf(buf, sizeof(buf), "Got pipeline key (%u bytes)\n", static_cast<unsigned>(key.keySize));
-		log_text_ += buf;
+		std::string message = std::format("Got pipeline key ({} bytes)", key.keySize);
+		LOGI(message);
```
The `LOGI` above this section can be removed.

asuessenbach commented:
You might need to `#include <format>`.

gary-sweet commented:
I see a bunch of compilation errors:

```
samples/extensions/pipeline_binary/pipeline_binary.cpp: In member function 'void PipelineBinary::log_pipeline_binary_support()':
samples/extensions/pipeline_binary/pipeline_binary.cpp:195:36: error: 'format' is not a member of 'std'; did you mean 'fmt::v11::format'?
195 | std::string message = std::format("VK_KHR_pipeline_binary support: pipelineBinaries = {}", !!features.pipelineBinaries);
| ^~~~~~
In file included from /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/components/core/include/core/util/error.hpp:54,
from framework/common/error.h:24,
from framework/camera.h:20,
from framework/api_vulkan_sample.h:34,
from samples/extensions/pipeline_binary/pipeline_binary.h:20,
from samples/extensions/pipeline_binary/pipeline_binary.cpp:18:
third_party/fmt/include/fmt/format.h:4198:31: note: 'fmt::v11::format' declared here
4198 | FMT_NODISCARD FMT_INLINE auto format(format_string<T...> fmt, T&&... args)
| ^~~~~~
samples/extensions/pipeline_binary/pipeline_binary.cpp:199:24: error: 'format' is not a member of 'std'; did you mean 'fmt::v11::format'?
199 | message = std::format(
| ^~~~~~
third_party/fmt/include/fmt/format.h:4198:31: note: 'fmt::v11::format' declared here
4198 | FMT_NODISCARD FMT_INLINE auto format(format_string<T...> fmt, T&&... args)
| ^~~~~~
samples/extensions/pipeline_binary/pipeline_binary.cpp: In member function 'void PipelineBinary::demo_pipeline_key_and_binary()':
samples/extensions/pipeline_binary/pipeline_binary.cpp:224:44: error: 'format' is not a member of 'std'; did you mean 'fmt::v11::format'?
224 | std::string message = std::format("vkGetPipelineKeyKHR failed ({}); skipping binary capture", vk::to_string(static_cast<vk::Result>(res)));
| ^~~~~~
third_party/fmt/include/fmt/format.h:4198:31: note: 'fmt::v11::format' declared here
4198 | FMT_NODISCARD FMT_INLINE auto format(format_string<T...> fmt, T&&... args)
| ^~~~~~
samples/extensions/pipeline_binary/pipeline_binary.cpp:232:44: error: 'format' is not a member of 'std'; did you mean 'fmt::v11::format'?
232 | std::string message = std::format("Got pipeline key ({} bytes)", key.keySize);
| ^~~~~~
third_party/fmt/include/fmt/format.h:4198:31: note: 'fmt::v11::format' declared here
4198 | FMT_NODISCARD FMT_INLINE auto format(format_string<T...> fmt, T&&... args)
| ^~~~~~
samples/extensions/pipeline_binary/pipeline_binary.cpp:251:44: error: 'format' is not a member of 'std'; did you mean 'fmt::v11::format'?
251 | std::string message = std::format("vkCreatePipelineBinariesKHR failed ({}); driver may not support capturing binaries in this context",
| ^~~~~~
third_party/fmt/include/fmt/format.h:4198:31: note: 'fmt::v11::format' declared here
4198 | FMT_NODISCARD FMT_INLINE auto format(format_string<T...> fmt, T&&... args)
| ^~~~~~
samples/extensions/pipeline_binary/pipeline_binary.cpp:267:44: error: 'format' is not a member of 'std'; did you mean 'fmt::v11::format'?
267 | std::string message = std::format("vkGetPipelineBinaryDataKHR size query failed ({}); skipping data fetch", vk::to_string(static_cast<vk::Result>(res)));
| ^~~~~~
third_party/fmt/include/fmt/format.h:4198:31: note: 'fmt::v11::format' declared here
4198 | FMT_NODISCARD FMT_INLINE auto format(format_string<T...> fmt, T&&... args)
| ^~~~~~
samples/extensions/pipeline_binary/pipeline_binary.cpp:287:52: error: 'format' is not a member of 'std'; did you mean 'fmt::v11::format'?
287 | std::string message = std::format("Binary signature: {} {} {} {} ...", binary_data_[0], binary_data_[1], binary_data_[2], binary_data_[3]);
| ^~~~~~
third_party/fmt/include/fmt/format.h:4198:31: note: 'fmt::v11::format' declared here
4198 | FMT_NODISCARD FMT_INLINE auto format(format_string<T...> fmt, T&&... args)
| ^~~~~~
samples/extensions/pipeline_binary/pipeline_binary.cpp:296:44: error: 'format' is not a member of 'std'; did you mean 'fmt::v11::format'?
296 | std::string message = std::format("vkGetPipelineBinaryDataKHR failed ({}); data not available", vk::to_string(static_cast<vk::Result>(res)));
| ^~~~~~
third_party/fmt/include/fmt/format.h:4198:31: note: 'fmt::v11::format' declared here
4198 | FMT_NODISCARD FMT_INLINE auto format(format_string<T...> fmt, T&&... args)
| ^~~~~~
```

gpx1000 committed:
Add missing std::format include to pipeline_binary sample

gpx1000 committed:
Remove duplicate logging statement in pipeline_binary sample

gpx1000 commented:
hmm...  Issue here.  std format was introduced in C++20.  I don't think we've updated this project to require C++20 yet.  @asuessenbach are you okay if I revert back to the original version?

SaschaWillems commented:
We did update to C++20 in march 2025, see #1257

gpx1000 committed:
Revert updating volk.

gpx1000 commented:
ahh... right, I forgot thanks!  Okay, fixing CI to allow for format.

gpx1000 committed:
Add `-extra-arg=-std=c++20` to clang-tidy and fix include order

Configure clang-tidy to use C++20 standard and reformat command for readability. Fix include order in pipeline_binary.cpp to follow conventions (standard headers alphabetically).

gpx1000 committed:
Fix clang-tidy by installing GCC 13 toolchain and configuring gcc-toolchain path

Install g++-13 to provide libstdc++ headers and configure clang-tidy to use GCC 13 toolchain via --gcc-toolchain flag to resolve C++20 standard library compatibility issues.

gpx1000 committed:
Remove clang-tidy container and use system-installed run-clang-tidy

Remove Docker container dependency and install steps. Detect and use available run-clang-tidy version (18, 17, or default) from the system. Remove GCC 13 toolchain configuration as it's no longer needed.

gpx1000 committed:
Disable X11 in GLFW build for clang-tidy check

gpx1000 committed:
Install XCB headers for CMake configure and re-enable X11 in GLFW build

Install libxcb1-dev to satisfy CMake dependencies and remove GLFW_BUILD_X11=OFF flag to restore default X11 support.

gpx1000 committed:
Disable X11 in GLFW build for clang-tidy check

gpx1000 commented:
Alright, the clang-tidy check in CI finally passed.  I might recommend we take this PR before we take any other PRs that use C++20 so the CI has a chance to stay green.

gary-sweet commented:
I'm still seeing one build error sadly:

```
samples/extensions/pipeline_binary/pipeline_binary.cpp:26:10: fatal error: format: No such file or directory
26 | #include <format>
| ^~~~~~~~
```

gpx1000 commented:
That's what I was talking about with regards to do we want to use C++20 everywhere.  format is the C++20 header.  You might need to see if you can't get a newer STL.  clang-tidy had the same issues which is what I fixed with the CI check.yml file if you want to review it.

gary-sweet commented:
> That's what I was talking about with regards to do we want to use C++20 everywhere. format is the C++20 header. You might need to see if you can't get a newer STL. clang-tidy had the same issues which is what I fixed with the CI check.yml file if you want to review it.

Hmm. Well we (as in the GPU group) don't have any control over the toolchain that is required to be used for our set-top boxes unfortunately. We're currently forced to use gcc12. If that's the problem here we won't be able to build anything at all after this.

gpx1000 commented:
Let's discuss on the next Samples call.  We can targeted only rely upon things that also exist in gcc12 in my opinion as I don't think we get anything but syntactic sugar with std::format instead of what I had previously; also I really don't want to exclude any devices or people that are actively helping if we can help it :)

gary-sweet commented:
> Let's discuss on the next Samples call. We can targeted only rely upon things that also exist in gcc12 in my opinion as I don't think we get anything but syntactic sugar with std::format instead of what I had previously; also I really don't want to exclude any devices or people that are actively helping if we can help it :)

Agreed. We literally wouldn't be able to run anything if we're forced down that route.

gpx1000 commented:
For this sample, I'll go ahead and revert the std format requirement (if that's okay @asuessenbach).  I'll do that in the morning though, it's my 1am.

gary-sweet commented:
> For this sample, I'll go ahead and revert the std format requirement (if that's okay @asuessenbach). I'll do that in the morning though, it's my 1am.

My advice - go to bed!

asuessenbach commented:
I think, I don't get that: we're on C++20 and gcc12 supports that nicely. But it doesn't have that `format` header?

asuessenbach commented:
> I'll go ahead and revert the std format requirement (if that's okay @asuessenbach).

I really don't like that `snprintf`-technique. But maybe that's just me.
Maybe you could as well just concatenate `std::string`s?
Like
`std::string message = std::string( "VK_KHR_pipeline_binary support: pipelineBinaries = " ) + features.pipelineBinaries ? "true" : "false";`

gary-sweet commented:
> I think, I don't get that: we're on C++20 and gcc12 supports that nicely. But it doesn't have that `format` header?

This is what AI has to say about it (believe as much of it as you dare):

```
While GCC 12 supports many C++20 features, it does not have a full implementation of <format>.

Full support for the std::format library in GCC didn't arrive until GCC 13. If you try to include <format> in GCC 12, you will likely run into a "file not found" error or find that the std::format namespace is missing.
```

SaschaWillems commented:
We do include [fmt](https://github.com/fmtlib/fmt) as a formatting library that's already used in the framework and samples. Why not use that instead of std::format?

gpx1000 committed:
Replace std::format with fmt::format in pipeline_binary sample

Switch from C++20 std::format to fmt library for broader compiler compatibility.
Fix formatting and indentation inconsistencies introduced in previous refactoring.

gpx1000 commented:
Excellent suggestion Sascha.  fmt seems to be a good direct replacement.

gpx1000 committed:
clang format fix.

gary-sweet commented:
More build errors I'm afraid:

```
samples/extensions/pipeline_binary/pipeline_binary.cpp: In member function 'void PipelineBinary::demo_pipeline_key_and_binary()':
samples/extensions/pipeline_binary/pipeline_binary.cpp:296:44: error: 'format' is not a member of 'std'; did you mean 'fmt::v11::format'?
296 | std::string message = std::format("vkGetPipelineBinaryDataKHR failed ({}); data not available", vk::to_string(static_cast<vk::Result>(res)));
| ^~~~~~
In file included from /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/components/core/include/core/util/error.hpp:54,
from framework/common/error.h:24,
from framework/camera.h:20,
from framework/api_vulkan_sample.h:34,
from samples/extensions/pipeline_binary/pipeline_binary.h:20,
from samples/extensions/pipeline_binary/pipeline_binary.cpp:18:
third_party/fmt/include/fmt/format.h:4198:31: note: 'fmt::v11::format' declared here
4198 | FMT_NODISCARD FMT_INLINE auto format(format_string<T...> fmt, T&&... args)
| ^~~~~~
```

gpx1000 commented:
Ah... I missed one...  simply replacing std to fmt will fix it.  I'll push a fix just a sec.

gpx1000 committed:
Replace std::format with fmt::format in pipeline_binary sample

gary-sweet commented:
Thanks, this does now build for me, and correctly reports that we don't support the extension.

gpx1000 committed:
Merge branch 'main' into VK_KHR_pipeline_binaries

SaschaWillems reviewed (changes requested):
Sample works fine, with one exception: Recreate pipeline (from binary) causes a validation error:

```
[error] 563254054 - VUID-vkCreateComputePipelines-pNext-09616: vkCreateComputePipelines(): pCreateInfos[0].pNext<VkPipelineBinaryInfoKHR>.binaryCount (1) is greater than zero while pipelineCache is not VK_NULL_HANDLE.
The Vulkan spec states: If VkPipelineBinaryInfoKHR::binaryCount is not 0 for any element of pCreateInfos, pipelineCache must be VK_NULL_HANDLE (https://docs.vulkan.org/spec/latest/chapters/pipelines.html#VUID-vkCreateComputePipelines-pNext-09616)
```

gpx1000 committed:
Merge branch 'main' into VK_KHR_pipeline_binaries

# Conflicts:
#	framework/vulkan_type_mapping.h
#	samples/extensions/README.adoc

gpx1000 committed:
address VUID error.

gpx1000 committed:
fix compile errors.

asuessenbach commented on .github/workflows/check.yml:
Are the changes to this file still needed?

gpx1000 replied:
No longer needed.  The updates are useful in the future but we can hold off for now.

gpx1000 committed:
Revert changes to clang-tidy check to main branch.  The updates are useful but not necessary for this PR.

gpx1000 committed:
Merge branch 'main' into VK_KHR_pipeline_binaries

asuessenbach commented:
Interestingly, recreating the pipeline from scratch is a bit faster on my machine than recreating it from binary.
Can anybody confirm this observation? Is that expected?

SaschaWillems commented:
I can imagine that this is caused by the pipeline used here being very light-weight and drivers using their own cache when not loading the binary. Result might change if you clear the driver cache before running the sample.

gpx1000 commented:
I think Sascha's reply matches my expectations.  Simple pipelines should be fast to just plain create right?  I think it'd need to be a larger or more complex pipeline to actually test performance.  Profiling seems slightly out of scope for a how-to sample though ;-)
